### DISCUSSION TITLE
Day 2

### DISCUSSION THREAD

**Discussion by AlecS19** [AlecS19/AoC-2024#2] [General]
**SPOILERS**

Discussion for Day 2 solutions

**Comment by AlecS19** [comment]
I messed up on day 2 part 2. 

Part 1 was easy, but took a little while for me to read and comprehend. Again basically utilized numpy arrays with logical statement to check if any difference was too big, or if it was monotonic.
`np.sum(np.abs(np.diff(arr)) > 3)` to check if gaps were too large
`np.sum(arr_diff < 0) == len(arr_diff)` to check for being monotonic

Part 2 I tried adapting my approach by tracking the number of errors... too complicated.
The better way was to make a function from part 1. If the original entry did not work, then I would remove one element and try again, quiting early if I found a successful entry.
```
if check(entry_arr):
        total += 1
        continue
    entry_arr = [int(i) for i in entry.split()]
    for i in range(len(entry_arr)):
        new_entry = copy(entry_arr)
        new_entry.pop(i)
        if check(new_entry):
            total +=1
            break
```
A bug that caused 5 minutes was `new_entry = entry`, which since they are lists are just references so I was not changing the array 1-by-1.
Solution was to use the `copy` functions

**Comment by AlecS19** [comment]
Pure python is just crazy to me
https://www.reddit.com/r/adventofcode/comments/1h4ncyr/comment/m0041k3/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

```python
def is_safe(row):
    inc = [row[i + 1] - row[i] for i in range(len(row) - 1)]
    if set(inc) <= {1, 2, 3} or set(inc) <= {-1, -2, -3}:
        return True
    return False

data = [[int(y) for y in x.split(' ')] for x in open('02.txt').read().split('\n')]

safe_count = sum([is_safe(row) for row in data])
print(safe_count)

safe_count = sum([any([is_safe(row[:i] + row[i + 1:]) for i in range(len(row))]) for row in data])
print(safe_count)
```

**Comment by jaxmichalski** [comment]
I also did it the same way in the end (except I didn't bother to check the whole array as my break statement would activate after one iteration of the list anyways). The issue I ran into was trying to do it logically instead of just brute forcing the permutations. Still not sure why I'm off but my "logic" way is still off by 2
